Allow Book to be used without specifying column types (and allow Fill to be called from python)

bendavid opened the pull request:
Allow Book to be used with RDataFrame without explicitly specifying the column types.  This re-uses as much as possible the machinery which was already in place to do this for Fill, Histo1D, etc

Template overloads are also simplified for Fill, which allows the jitted version at least to be called from python.

eguiraud commented:
@phsft-bot build

eguiraud commented:
@bendavid the warnings are real -- I will try to do a proper review tomorrow, busy days sorry

bendavid commented:
Ok yes I'll fix those warnings in the meantime.

eguiraud commented:
Thanks! At first glance it looks fine anyway, and especially Python users will be very glad to have a jitted version of Book

pcanal commented:
Looks nice.  Could we also add some test/tutorials?

bendavid commented:
Hmm, actually it seems that I cannot call Book directly from python with the overloads set up this way.

I think this has to do with reference type ambiguities when passing objects from python.

Let me try further further restricting the template instances with SFINAE  (I guess the same issue is present with Fill since it has the same template argument pattern)

```
Info in <TUnixSystem::ACLiC>: creating shared library /data/home/bendavid/rdftest7/jittest_cpp.so
input_line_45:7:122: error: call to member function 'Book' is ambiguous
         new (ret) (ROOT::RDF::RResultPtr<double>) (((ROOT::RDF::RInterface<ROOT::Detail::RDF::RLoopManager,void>*)obj)->Book<TestHelper>((TestHelper&&)*(TestHelper*)args[0]));
                                                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~
/data/home/bendavid/rootdevfillopt/install/include/ROOT/RDF/RInterface.hxx:2240:42: note: candidate function [with FirstColumn = TestHelper, OtherColumns = <>, Helper = TestHelper]
   RResultPtr<typename Helper::Result_t> Book(Helper &&helper, const ColumnNames_t &columns = {})
                                         ^
/data/home/bendavid/rootdevfillopt/install/include/ROOT/RDF/RInterface.hxx:2264:42: note: candidate function [with Helper = TestHelper]
   RResultPtr<typename Helper::Result_t> Book(Helper &&helper, const ColumnNames_t &columns = {})
                                         ^
input_line_45:11:79: error: call to member function 'Book' is ambiguous
         ((ROOT::RDF::RInterface<ROOT::Detail::RDF::RLoopManager,void>*)obj)->Book<TestHelper>((TestHelper&&)*(TestHelper*)args[0]);
         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~
/data/home/bendavid/rootdevfillopt/install/include/ROOT/RDF/RInterface.hxx:2240:42: note: candidate function [with FirstColumn = TestHelper, OtherColumns = <>, Helper = TestHelper]
   RResultPtr<typename Helper::Result_t> Book(Helper &&helper, const ColumnNames_t &columns = {})
                                         ^
/data/home/bendavid/rootdevfillopt/install/include/ROOT/RDF/RInterface.hxx:2264:42: note: candidate function [with Helper = TestHelper]
   RResultPtr<typename Helper::Result_t> Book(Helper &&helper, const ColumnNames_t &columns = {})
                                         ^
input_line_45:17:122: error: call to member function 'Book' is ambiguous
         new (ret) (ROOT::RDF::RResultPtr<double>) (((ROOT::RDF::RInterface<ROOT::Detail::RDF::RLoopManager,void>*)obj)->Book<TestHelper>((TestHelper&&)*(TestHelper*)args[0], (const vector<string>&)*(const vector<string>*)args[1]));
                                                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~
/data/home/bendavid/rootdevfillopt/install/include/ROOT/RDF/RInterface.hxx:2240:42: note: candidate function [with FirstColumn = TestHelper, OtherColumns = <>, Helper = TestHelper]
   RResultPtr<typename Helper::Result_t> Book(Helper &&helper, const ColumnNames_t &columns = {})
                                         ^
/data/home/bendavid/rootdevfillopt/install/include/ROOT/RDF/RInterface.hxx:2264:42: note: candidate function [with Helper = TestHelper]
   RResultPtr<typename Helper::Result_t> Book(Helper &&helper, const ColumnNames_t &columns = {})
                                         ^
input_line_45:21:79: error: call to member function 'Book' is ambiguous
         ((ROOT::RDF::RInterface<ROOT::Detail::RDF::RLoopManager,void>*)obj)->Book<TestHelper>((TestHelper&&)*(TestHelper*)args[0], (const vector<string>&)*(const vector<string>*)args[1]);
         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~
/data/home/bendavid/rootdevfillopt/install/include/ROOT/RDF/RInterface.hxx:2240:42: note: candidate function [with FirstColumn = TestHelper, OtherColumns = <>, Helper = TestHelper]
   RResultPtr<typename Helper::Result_t> Book(Helper &&helper, const ColumnNames_t &columns = {})
                                         ^
/data/home/bendavid/rootdevfillopt/install/include/ROOT/RDF/RInterface.hxx:2264:42: note: candidate function [with Helper = TestHelper]
   RResultPtr<typename Helper::Result_t> Book(Helper &&helper, const ColumnNames_t &columns = {})
                                         ^
Traceback (most recent call last):
  File "jittest.py", line 53, in <module>
    testfunc(book4)
  File "jittest.py", line 14, in testfunc
    res = f(df, "dentry")
  File "jittest.py", line 39, in book4
    res = rdf.Book(ROOT.TestHelper(nslots), [col])
TypeError: Template method resolution failed:
  Failed to instantiate "Book(TestHelper&,std::initializer_list<std::string>)"
  Failed to instantiate "Book(TestHelper*,std::initializer_list<std::string>)"
  ROOT::RDF::RResultPtr<double> ROOT::RDF::RInterface<ROOT::Detail::RDF::RLoopManager,void>::Book(TestHelper&& helper, const vector<string>& columns = {}) =>
    ValueError: nullptr result where temporary expected
```

bendavid commented:
Fixed the unused variable warnings, and simplified the template overloads so that Book can be called directly from python.  (At least in the jitted case.  I'm not sure how to call a template function like this from cppyy with partially specified template parameters, and it's not possible to explicitly specify the Helper type because it comes after the parameter pack.)

eguiraud reviewed (changes requested):
Looks good to me, it's just missing a test. Please add one in `tree/dataframe/test/dataframe_simple.cxx`, right after the test for the templated `Book` version.

And thank you very much for the nice contribution!

eguiraud commented:
Ah, about calling `Fill` from Python: see https://sft.its.cern.ch/jira/browse/ROOT-10396 , it's currently not possible -- please share any insight into that issue you might have gotten from this work on `Book` :smile:

bendavid commented:
One restriction I realized when adding a test is that the jitted version of Book will only work if the helper class is registered with the Root type system (because of the ```TClass::GetClass``` call at https://github.com/root-project/root/blob/d03407e19eaea7b02021d9d65ec6bc864bdcab6b/tree/dataframe/src/RDFInterfaceUtils.cxx#L680 )

Open to suggestions how to remove this restriction (demangling the name with ```abi::_cxa_demangle``` maybe?)  Otherwise I can add a comment about this to the documentation for now.

bendavid commented on tree/dataframe/inc/ROOT/RDF/RInterface.hxx:
```diff
@@ -2397,7 +2389,7 @@ private:
              typename std::enable_if<!RDFInternal::RNeedJitting<ColTypes...>::value, int>::type = 0>
    RResultPtr<ActionResultType> CreateAction(const ColumnNames_t &columns, const std::shared_ptr<ActionResultType> &r,
-                                             const std::shared_ptr<HelperArgType> &helperArg)
+                                             const std::shared_ptr<HelperArgType> &helperArg, const int)
```
Done (and also fixed a missing default value which was needed here)

bendavid commented:
I added the corresponding change to template overloads for Fill directly to this PR (and also added the additional test as requested)

eguiraud reviewed (approved):
LGTM, please mention ROOT-10396 in the message of the commit that fixes it (EDIT: and thank you very much for fixing it in the first place!). Other than that, if jenkins agrees we can merge this.

eguiraud commented:
@phsft-bot build

eguiraud commented:
> Open to suggestions how to remove this restriction (demangling the name with abi::_cxa_demangle maybe?) Otherwise I can add a comment about this to the documentation for now.

I don't think we can lift that restriction, cling needs to jit code that makes use of the class, so it needs to know about the class. Indeed please add a note about this to the `Book` docstring.

bendavid commented:
Updated/fixed the docstrings and added ROOT-10396 to the commit message as discussed.

eguiraud commented:
Thanks Josh! Unfortunately the test failures above seem to be caused by the changes in this PR but I have no idea why or how to fix them (yet). Anyway it's on us.

These changes are ok, it's on my to do list to debug the pyroot failures, given my current bandwidth it will probably have to be after Easter :/ sorry

eguiraud commented:
@phsft-bot build please

eguiraud commented:
With @etejedor 's help we could reduce the problem that causes the test failures to:

```cpp
#include "TClassRef.h"

int main() {
   TClassRef cr("ROOT::RDF::RInterface<ROOT::Detail::RDF::RLoopManager,void>");
   cr->GetListOfMethods(true);
   return 0;
}
```

which breaks with the following stacktrace:

```
#0  0x00007ffff53f9ef5 in raise () from /usr/lib/libc.so.6
#1  0x00007ffff53e3862 in abort () from /usr/lib/libc.so.6
#2  0x00007ffff53e3747 in __assert_fail_base.cold () from /usr/lib/libc.so.6
#3  0x00007ffff53f2646 in __assert_fail () from /usr/lib/libc.so.6
#4  0x00007fffec0bcec0 in TClingCXXRecMethIter::InstantiateTemplateWithDefaults (this=0x555556ead490, TD=0x555556be3a30) at ../../../../bendavidbook/core/metacling/src/TClingMethodInfo.cxx:182
#5  0x00007fffec0ba7d8 in TClingMemberIter::Advance (this=0x555556ead490) at ../../../../bendavidbook/core/metacling/src/TClingMemberIter.cxx:140
#6  0x00007fffec0b8b0f in TClingMemberIter::Next (this=0x555556ead490) at ../../../../bendavidbook/core/metacling/src/TClingMemberIter.h:159
#7  0x00007fffec0be2cd in TClingMethodInfo::Next (this=0x555556ead450) at ../../../../bendavidbook/core/metacling/src/TClingMethodInfo.cxx:438
#8  0x00007fffebf51dba in TCling::MethodInfo_Next (this=0x5555556433a0, minfo=0x555556ead450) at ../../../../bendavidbook/core/metacling/src/TCling.cxx:9017
#9  0x00007ffff7d8192f in TListOfFunctions::Load (this=0x555556dbb2b0) at ../../../bendavidbook/core/meta/src/TListOfFunctions.cxx:391
#10 0x00007ffff7d51b82 in TClass::GetListOfMethods (this=0x5555561abdb0, load=true) at ../../../bendavidbook/core/meta/src/TClass.cxx:3803
#11 0x00005555555551dd in main () at repro.cpp:5
```

Work in progress.

eguiraud commented:
@bendavid just fyi, #7898 by @Axel-Naumann should fix these test failures, it's a bug in cling. We should be able to merge this PR then.

eguiraud commented:
Reopening, stupid github

eguiraud commented:
The Windows failure is unrelated, merging. Thanks again @bendavid !

eguiraud commented:
@bendavid I realized just now that this breaks the case of an action helper that takes no arguments, or it forces jitting in that case. I'll try to think of a way to have our cake and eat it too, but if you have any suggestions... :smile:

eguiraud commented:
Update: this is what I could come up with to resolve the ambiguity: https://github.com/root-project/root/commit/08ca2bbd6dd195daa1d6f635f6a5e8fed5ef4fd4 -- will propose a PR after adding a test
